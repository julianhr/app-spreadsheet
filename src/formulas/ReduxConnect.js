import appStore from '~/reducers/'
import { clearCellData, replaceCellData } from '~/actions/tableActions'


class ReduxConnect {
  constructor(store=appStore) {
    const { global: globalState } = store.getState()
    this.store = store
    this.rows = globalState.rows
    this.columns = globalState.columns
  }

  get table() {
    return this.store.getState().table
  }

  get locations() {
    return Object.keys(this.table)
  }

  getCellResult(location) {
    const data = this.table[location]
    return data && data.result
  }

  replaceCellData(location, entered, result) {
    this.store.dispatch(replaceCellData(location, entered, result))
  }

  clearCell(location) {
    this.store.dispatch(clearCellData(location))
  }
}

export default ReduxConnect
export { appStore }
