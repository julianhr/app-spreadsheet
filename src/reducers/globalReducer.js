import { createReducer } from 'redux-starter-kit'

export const INITIAL_STATE = {
  activeCell: null,
  rows: 14,
  columns: 6,
}

export default createReducer(INITIAL_STATE, {
  'SET_ACTIVE_CELL': (state, { payload }) => {
    state.activeCell = payload
  },
})
