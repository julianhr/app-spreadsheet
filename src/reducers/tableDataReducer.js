import { createReducer } from 'redux-starter-kit'


const INITIAL_STATE = {}

export default createReducer(INITIAL_STATE, {
  'SET_CELL_DATA': (state, { payload }) => {
    const { location, data } = payload
    state[location] = data
  },

  'DELETE_CELL_DATA': (state, { payload }) => {
    delete state[payload]
  }
})
