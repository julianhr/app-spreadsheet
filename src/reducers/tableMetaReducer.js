import { createReducer } from 'redux-starter-kit'
import { DEFAULT_COL_WIDTH } from '~/library/constants'


const INITIAL_STATE = {
  colWidths: {},
}

export default createReducer(INITIAL_STATE, {
  'SET_COL_WIDTH': (state, { payload }) => {
    const { colLabel, width } = payload
    const { colWidths } = state
    
    if (width === DEFAULT_COL_WIDTH) {
      delete state.colWidths[colLabel]
    } else {
      state.colWidths = { ...colWidths, [colLabel]: width }
    }
  },
})
