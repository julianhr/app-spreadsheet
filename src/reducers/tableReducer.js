import { createReducer } from 'redux-starter-kit'


const INITIAL_STATE = {}

export default createReducer(INITIAL_STATE, {
  'SET_CELL_VALUE': (state, { payload }) => {
    const { location, text, formula } = payload
    state[location] = { text, formula }
  },

  'DELETE_CELL_VALUE': (state, { payload }) => {
    delete state[payload]
  }
})
