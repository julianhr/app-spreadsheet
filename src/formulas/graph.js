import ReduxConnect from './ReduxConnect'
import Interpreter from './Interpreter'


class Vertex {
  constructor(entered) {
    this.entered = entered
    this.error = null
    this.result = null
    this.ast = null
  }
}

class Graph {
  constructor() {
    this.adj = {}
    this.pending = null
    this.visited = null
    this.store = new ReduxConnect()
  }

  recalculate(location) {
    this.setPendingNodes()
    this.dfs(location)
    this.updateStore()
  }

  setPendingNodes() {
    this.pending = new Set()

    Object.keys(this.adj).forEach(vertex => {
      this.pending.add(vertex)
    })
  }

  dfs(location) {
    while (this.pending.size > 0) {
      if (!location) {
        location = [...this.pending][0]
      }

      try {
        this.visited = new Set([location])
        const interpreter = new Interpreter(location)
        interpreter.visitAST(location)
      } catch (error) {}

      this.visited.forEach(location => this.pending.delete(location))
      location = null
    }

    this.visited = null
    this.pending = null
  }

  visitCell(location) {
    if (!this.hasVertex(location)) {
      return 0
    }

    this.testPathCycle(location)
    this.pending.delete(location)
    this.visited.add(location)
    const interpreter = new Interpreter(location)
    const result = interpreter.visitAST(location)
    this.visited.delete(location)
    return result
  }

  getCellResult(location) {
    if (location in this.adj) {
      return graph.adj[location].result
    } else {
      return ''
    }
  }

  testPathCycle(location) {
    if (this.visited.has(location)) {
      throw new Error('Circular reference')
    }
  }

  updateStore() {
    const locationKeys = new Set(this.store.locations)

    for (let [location, vertex] of Object.entries(this.adj)) {
      const { entered, result } = vertex

      locationKeys.delete(location)

      if (result !== this.store.getCellResult(location)) {
        this.store.replaceCellData(location, entered, result)
      }
    }

    for (let location of locationKeys) {
      this.store.clearCell(location)
    }
  }

  addVertex(location, entered) {
    const vertex = new Vertex(entered)
    this.adj[location] = vertex
    return vertex
  }

  hasVertex(location) {
    return location in this.adj
  }

  getVertex(location) {
    this.testMissingLocation(location)
    return this.adj[location]
  }

  delVertex(location) {
    if (location in this.adj) {
      delete this.adj[location]
      this.recalculate()
    }
  }

  resetAll() {
    this.adj = {}
    this.pending = null
    this.visited = null
    this.store = new ReduxConnect()
  }

  testMissingLocation(location) {
    if (!this.hasVertex(location)) {
      throw new Error('Missing location')
    }
  }
}

const graph = new Graph()


export default graph
export {
  Vertex,
  Graph,
}
