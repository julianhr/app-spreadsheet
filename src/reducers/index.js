import { configureStore } from 'redux-starter-kit'

import globalReducer from './globalReducer'
import tableReducer from './tableReducer'


const appStoreFn = () => (
  configureStore({
    reducer: {
      global: globalReducer,
      table: tableReducer,
    }
  })
)


export default appStoreFn
