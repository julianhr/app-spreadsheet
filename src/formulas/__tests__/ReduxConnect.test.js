import graph from '../graph'
import { appStoreGen } from '~/reducers/'
import ReduxConnect from '../ReduxConnect'


describe('ReduxConnect', () => {
  let rc

  beforeEach(() => {
    rc = new ReduxConnect(appStoreGen())
  })

  test('instantiation', () => {
    expect(rc).toMatchSnapshot()
  })

  test('get #table', () => {
    rc.replaceCellData('A-1', '7', 7)
    rc.replaceCellData('B-2', '=A1+3', 10)
    expect(rc.table).toMatchSnapshot()
  })

  test('get #locations', () => {
    rc.replaceCellData('A-1', '7', 7)
    rc.replaceCellData('B-2', '=A1+3', 10)
    expect(rc.locations).toEqual(['A-1', 'B-2'])
  })

  test('#getCellResult', () => {
    rc.replaceCellData('B-2', '=A1+3', 3)
    expect(rc.getCellResult('B-2')).toBe(3)
  })

  test('#replaceCellData', () => {
    rc.replaceCellData('B-2', '=A1+3', 3)
    expect(rc.table).toMatchSnapshot()

    rc.replaceCellData('B-2', '=c5/7', 0)
    expect(rc.table).toMatchSnapshot()
  })

  test('#clearCell', () => {
    rc.replaceCellData('B-2', '=A1+3', 3)
    expect(rc.getCellResult('B-2')).toBe(3)

    rc.clearCell('B-2')
    expect(rc.getCellResult('B-2')).toBeUndefined()
  })
})
