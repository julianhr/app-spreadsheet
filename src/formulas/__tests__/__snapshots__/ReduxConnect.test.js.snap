// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReduxConnect #replaceCellData 1`] = `
Object {
  "B-2": Object {
    "entered": "=A1+3",
    "result": 3,
  },
}
`;

exports[`ReduxConnect #replaceCellData 2`] = `
Object {
  "B-2": Object {
    "entered": "=c5/7",
    "result": 0,
  },
}
`;

exports[`ReduxConnect get #table 1`] = `
Object {
  "A-1": Object {
    "entered": "7",
    "result": 7,
  },
  "B-2": Object {
    "entered": "=A1+3",
    "result": 10,
  },
}
`;

exports[`ReduxConnect instantiation 1`] = `
ReduxConnect {
  "columns": 6,
  "rows": 14,
  "store": Object {
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
    Symbol(observable): [Function],
  },
}
`;
